### ISSUE TITLE
[Refactor] Create loss_utils.mojo for shared loss function patterns

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2468]
## Problem

`shared/core/loss.mojo` (931 lines) has 14 loss functions sharing similar patterns:

**Loss functions:**
- Binary Cross-Entropy (forward + backward)
- Mean Squared Error (forward + backward)
- Cross-Entropy (forward + backward)
- Smooth L1 Loss (forward + backward)
- Hinge Loss (forward + backward)
- Focal Loss (forward + backward)
- KL Divergence (forward + backward)

**Duplicated patterns:**
1. **Input validation** (~30 lines total):
   - Shape matching check
   - Dtype validation
   - Range validation for probabilities

2. **Epsilon handling** (~50 lines total):
   - Clipping predictions to [epsilon, 1-epsilon]
   - Adding epsilon for log safety

3. **One-tensor creation** (~20 lines total):
   - Creating tensors of ones for broadcasting
   - Creating epsilon tensors

## Files Affected

- `shared/core/loss.mojo` (931 lines)
- New: `shared/core/loss_utils.mojo`

## Proposed Solution

1. Create `shared/core/loss_utils.mojo`:
   ```mojo
   fn validate_loss_inputs(
       predictions: ExTensor,
       targets: ExTensor
   ) raises:
       """Validate prediction and target tensors for loss computation."""
       if predictions.shape() != targets.shape():
           raise Error("Predictions and targets must have same shape")
       if predictions.dtype() != targets.dtype():
           raise Error("Predictions and targets must have same dtype")
   
   fn clip_predictions(
       predictions: ExTensor,
       epsilon: Float64 = 1e-7
   ) -> ExTensor:
       """Clip predictions to [epsilon, 1-epsilon] for numerical stability."""
       var eps_tensor = full_like(predictions, epsilon)
       var one_minus_eps = full_like(predictions, 1.0 - epsilon)
       return maximum(minimum(predictions, one_minus_eps), eps_tensor)
   
   fn apply_epsilon_safety(
       values: ExTensor,
       epsilon: Float64 = 1e-7
   ) -> ExTensor:
       """Add epsilon to values for log safety."""
       var eps_tensor = full_like(values, epsilon)
       return add_simd(values, eps_tensor)
   
   fn create_weight_tensor(
       reference: ExTensor,
       weights: Optional[ExTensor] = None
   ) -> ExTensor:
       """Create weight tensor or default to ones."""
       if weights:
           return weights.value()
       return ones_like(reference)
   ```

2. Refactor loss functions to use shared utilities:
   ```mojo
   fn binary_cross_entropy_forward(
       predictions: ExTensor,
       targets: ExTensor,
       epsilon: Float64 = 1e-7
   ) -> ExTensor:
       validate_loss_inputs(predictions, targets)
       var safe_pred = clip_predictions(predictions, epsilon)
       # ... BCE computation
   ```

## Expected Impact

- **~230 lines saved (25% reduction)**
- Consistent numerical stability
- Easier to add new loss functions
- Centralized epsilon handling

## Priority

Medium (Phase 3)
